hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java

```diff
@@ -270,13 +270,36 @@ private DistributionSetAssignmentResult assignDistributionSetToTargets(final Lon
                 distributionSetEntity.getId());
 
         if (targetEntities.isEmpty()) {
-            // detaching as it is not necessary to persist the set itself
-            entityManager.detach(distributionSetEntity);
-            // return with nothing as all targets had the DS already assigned
-            return new DistributionSetAssignmentResult(distributionSetEntity, Collections.emptyList(), 0,
-                    targetsWithActionType.size(), Collections.emptyList(), targetManagement);
+            return allTargetsAlreadyAssignedResult(distributionSetEntity, targetsWithActionType.size());
         }
 
+        closeOrCancelActiveActionsAndSetAssignedDSAndTargetUpdateStatus(assignmentStrategy, distributionSetEntity,
+                targetEntities);
+
+        final Map<String, JpaAction> controllerIdsToActions = createActions(targetsWithActionType, targetEntities,
+                assignmentStrategy, distributionSetEntity);
+        // create initial action status when action is created so we remember
+        // the initial running status because we will change the status
+        // of the action itself and with this action status we have a nicer
+        // action history.
+        createActionsStatus(controllerIdsToActions.values(), assignmentStrategy, actionMessage);
+
+        detachEntitiesAndSendTargetUpdatedEvents(distributionSetEntity, targetEntities, assignmentStrategy);
+        return buildAssignmentResult(distributionSetEntity, controllerIdsToActions);
+    }
+
+    private DistributionSetAssignmentResult allTargetsAlreadyAssignedResult(
+            final JpaDistributionSet distributionSetEntity, final int alreadyAssignedCount) {
+        // detaching as it is not necessary to persist the set itself
+        entityManager.detach(distributionSetEntity);
+        // return with nothing as all targets had the DS already assigned
+        return new DistributionSetAssignmentResult(distributionSetEntity, alreadyAssignedCount, Collections.emptyList(),
+                Collections.emptyList(), Collections.emptyList());
+    }
+
+    private void closeOrCancelActiveActionsAndSetAssignedDSAndTargetUpdateStatus(
```

Comment:
try to find a better name to describe why this method is called

Reply:
*And* in the middle of the method name is considered to be bad :) (single responsibility principle)

Reply:
This method `assignDistributionSetToTargets` is anyway doing too many things, this is why I wanted to refactor it. It is challenging to find suitable names :)

Reply:
You haven't seen it before =) Still, I would suggest dividing it into two separate methods, because then the semantic is clear, and yes, there are many steps, but the steps themselves are atomic